File: Baekjoon/Class3/47.9019 DSLR.js
// 양방향 BFS
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const MAX_NUM = 9999;
const DIGIT_LIMIT = 10000;
const BEGINNING = -1;
const A_OPERATIONS = "DSLR";
const B_OPERATIONS = "DDSLR";

// 미리 DSLR 계산
const A_DSLR = Array.from({ length : MAX_NUM + 1 }, (_, n) => [
    (2 * n) % DIGIT_LIMIT,
    n == 0 ? MAX_NUM : n - 1,
    (n * 10) % DIGIT_LIMIT + Math.trunc(n / (DIGIT_LIMIT / 10)), // shift left + carry
    Math.trunc((n / 10) + (n % 10) * (DIGIT_LIMIT / 10)) // shift right + carry * 1000
]);
const B_DSLR = Array.from({ length : MAX_NUM + 1 }, (_, n) => [
    Math.trunc(n / 2),
    Math.trunc((n + DIGIT_LIMIT) / 2),
    n == MAX_NUM ? 0 : n + 1,
    Math.trunc((n / 10) + (n % 10) * (DIGIT_LIMIT / 10)), // shift right + carry * 1000
    (n * 10) % DIGIT_LIMIT + Math.trunc(n / (DIGIT_LIMIT / 10)) // shift left + carry
]);

const getTraces = (meetingPoint, aOperations, bOperations) => {
    const trace = []
    let cur = aOperations[meetingPoint]; // A 역추적
    while(cur[1] !== BEGINNING) {
        trace.push(cur[0]);
        cur = aOperations[cur[1]];
    }
    trace.reverse();
    
    cur = bOperations[meetingPoint]; // B 역추적
    while(cur[1] !== BEGINNING) {
        trace.push(cur[0]);
        cur = bOperations[cur[1]];
    }
    return trace.join("");
}

let cursor = 1;
const runRegister = (out) => {
    const [a, b] = input[cursor++].split(" ").map(Number);
    if(a === b) {
        out.push("");
        return;
    }

    const aOperations = new Array(MAX_NUM + 1).fill(null); // [operation, prevIndex]
    const bOperations = new Array(MAX_NUM + 1).fill(null);
    let aBfsHead = 0;
    let bBfsHead = 0;
    const aBfs = [a]; // 너비 우선 탐색
    const bBfs = [b];
    aOperations[a] = ["", BEGINNING];
    bOperations[b] = ["", BEGINNING];

    while(true) {
        let levelLimit = aBfs.length;
        while(aBfsHead < levelLimit) { // 한 레벨씩
            const start = aBfs[aBfsHead++];

            // 정방향 탐색
            for(let i = 0; i < 4; i++) {
                const end = A_DSLR[start][i];
                if(aOperations[end] === null) { // 연산자 저장
                    aBfs.push(end);
                    aOperations[end] = [A_OPERATIONS[i], start];
                }
                if(bOperations[end] !== null) { // 반대편과 만남
                    out.push(getTraces(end, aOperations, bOperations));
                    return;
                }
            }
        }

        levelLimit = bBfs.length;
        while(bBfsHead < levelLimit) {
            const start = bBfs[bBfsHead++];

            // 정방향 탐색
            for(let i = (start % 2 === 0) ? 0 : 2; i < 5; i++) {
                const end = B_DSLR[start][i];
                if(bOperations[end] === null) { // 연산자 저장
                    bBfs.push(end);
                    bOperations[end] = [B_OPERATIONS[i], start];
                }
                if(aOperations[end] !== null) { // 반대편과 만남
                    out.push(getTraces(end, aOperations, bOperations));
                    return;
                }
            }
        }
    }
}

const t = Number(input[0]);
const out = [];
for(let _ = 0; _ < t; _++)
    runRegister(out)
console.log(out.join("\n"));

/* 단방향 BFS
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const MAX_NUM = 9999;
const DIGIT_LIMIT = 10000;
const BEGINNING = -1;
const OPERATIONS = "DSLR";

let cursor = 1;
const runRegister = (out) => {
    const [a, b] = input[cursor++].split(" ").map(Number);

    const operations = new Array(MAX_NUM + 1).fill(null); // [operation, prevIndex]
    let bfsHead = 0;
    const bfs = [a]; // 너비 우선 탐색
    operations[a] = ["", BEGINNING];
    while(true) {
        const start = bfs[bfsHead++];

        const ends = [ // DSLR 계산
            (2 * start) % DIGIT_LIMIT,
            start === 0 ? MAX_NUM : start - 1,
            (start * 10) % DIGIT_LIMIT + Math.trunc(start / (DIGIT_LIMIT / 10)), // shift left + carry
            Math.trunc((start / 10) + (start % 10) * (DIGIT_LIMIT / 10)) // shift right + carry * 1000
        ];

        for(let i = 0; i < ends.length; i++) {
            const end = ends[i];
            if(operations[end] === null) { // 연산자 저장
                bfs.push(end);
                operations[end] = [OPERATIONS[i], start];
            }
            if(end === b) { // B 도착
                const trace = []
                let cur = operations[end]; // 역추적
                while(cur[1] !== BEGINNING) {
                    trace.push(cur[0]);
                    cur = operations[cur[1]];
                }
                out.push(trace.reverse().join(""));
                return;
            }
        }
    }
}

const t = Number(input[0]);
const out = [];
for(let _ = 0; _ < t; _++)
    runRegister(out)
console.log(out.join("\n"));
*/

File: Baekjoon/Class3/47.9019 DSLR.py
#양방향 BFS
from collections import deque
import sys
input = sys.stdin.readline
write = sys.stdout.write

MAX_NUM = 9999
BEGINNING = -1
A_OPERATIONS = "DSLR"
B_OPERATIONS = "DDSLR"

#미리 DSLR 계산
A_DSLR = [(
    (2 * n) % 10000,
    MAX_NUM if n == 0 else n - 1,
    (n * 10) % 10000 + n // 1000, #shift left + carry
    (n // 10) + (n % 10) * 1000 #shift right + carry * 1000
    ) for n in range(10000)]
B_DSLR = [(
    n // 2,
    (n + 10000) // 2,
    0 if n == 9999 else n + 1,
    (n // 10) + (n % 10) * 1000, #shift right + carry * 1000
    (n * 10) % 10000 + n // 1000 #shift left + carry
    ) for n in range(10000)]

def print_traces(meeting_point, a_operations, b_operations):
    a_trace = []
    cur = a_operations[meeting_point] #A 역추적
    while cur[1] != BEGINNING:
        a_trace.append(cur[0])
        cur = a_operations[cur[1]]

    b_trace = []
    cur = b_operations[meeting_point] #B 역추적
    while cur[1] != BEGINNING:
        b_trace.append(cur[0])
        cur = b_operations[cur[1]]
    
    write(f"{"".join(a_trace[::-1])}{"".join(b_trace)}\n")

def find_meeting_point(start, ends, bfs, operations, other_operations):
    for symbol, end in ends:
        if operations[end] is None: #연산자 저장
            bfs.append(end)
            operations[end] = (symbol, start)
        if other_operations[end] is not None: #반대편과 만남
            return end
    return None

def run_register():
    a, b = map(int, input().split())
    if a == b:
        write("\n")
        return

    a_operations = [None] * (MAX_NUM + 1) #(operation, prev_index)
    b_operations = [None] * (MAX_NUM + 1)
    a_bfs = deque([a]) #너비 우선 탐색
    b_bfs = deque([b])
    a_operations[a] = ("", BEGINNING)
    b_operations[b] = ("", BEGINNING)

    while True:
        for _ in range(len(a_bfs)): #한 레벨씩
            #정방향 DSLR 연산 저장
            start = a_bfs.popleft()
            ends = [(A_OPERATIONS[i], A_DSLR[start][i]) for i in range(4)]

            #정방향 탐색
            meeting_point = find_meeting_point(start, ends, a_bfs, a_operations, b_operations)
            if meeting_point is not None:
                print_traces(meeting_point, a_operations, b_operations)
                return

        for _ in range(len(b_bfs)):
            #역방향 DSLR 연산 저장
            start = b_bfs.popleft()
            ends = [(B_OPERATIONS[i], B_DSLR[start][i]) for i in (range(0, 5) if start % 2 == 0 else range(2, 5))]

            #역방향 탐색
            meeting_point = find_meeting_point(start, ends, b_bfs, b_operations, a_operations)
            if meeting_point is not None:
                print_traces(meeting_point, a_operations, b_operations)
                return

def main():
    t = int(input())
    for _ in range(t):
        run_register()

main()

''' 단방향 BFS
from collections import deque
import sys
input = sys.stdin.readline
write = sys.stdout.write

MAX_NUM = 9999
NOT_VISITED = -1
BEGINNING = -2
OPERATIONS = "DSLR"

#미리 DSLR 계산
DSLR = [(
    (2 * n) % 10000,
    MAX_NUM if n == 0 else n - 1,
    (n * 10) % 10000 + n // 1000, #shift left + carry
    (n // 10) + (n % 10) * 1000 #shift right + carry * 1000
    ) for n in range(10000)]

def run_register():
    a, b = map(int, input().split())

    operations = [""] * (MAX_NUM + 1)
    prev_indexes = [NOT_VISITED] * (MAX_NUM + 1)
    bfs = deque([a]) #너비 우선 탐색
    prev_indexes[a] = BEGINNING
    while True:
        start = bfs.popleft()
        for i in range(len(OPERATIONS)):
            end = DSLR[start][i]
            if prev_indexes[end] == NOT_VISITED: #연산자 저장
                bfs.append(end)
                operations[end] = OPERATIONS[i]
                prev_indexes[end] = start
            if end == b: #B 도착
                out = [operations[end]]
                prev_index = prev_indexes[end] #역추적
                while prev_index != BEGINNING:
                    out.append(operations[prev_index])
                    prev_index = prev_indexes[prev_index]
                write(f"{"".join(out[::-1])}\n")
                return

def main():
    t = int(input())
    for _ in range(t):
        run_register()

main()
'''